go/internal/feast/onlinestore/cassandraonlinestore.go

```diff
@@ -0,0 +1,392 @@
+	}
+	results := make([][]FeatureData, len(entityKeys))
+	for i := range results {
+		results[i] = make([]FeatureData, len(featureNames))
+	}
+
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	featureViewName := featureViewNames[0]
+
+	// Prepare the query
+	tableName := c.getFqTableName(featureViewName)
+	cqlStatement := c.getCQLStatement(tableName, featureNames)
+
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(len(serializedEntityKeys))
+
+	errorsChannel := make(chan error, len(serializedEntityKeys))
+	for _, serializedEntityKey := range serializedEntityKeys {
+		go func(serEntityKey any) {
+			defer waitGroup.Done()
+
+			iter := c.session.Query(cqlStatement, serEntityKey).WithContext(ctx).Iter()
+
+			rowIdx := serializedEntityKeyToIndex[serializedEntityKey.(string)]
+
+			// fill the row with nulls if not found
+			if iter.NumRows() == 0 {
+				for _, featName := range featureNames {
+					results[rowIdx][featureNamesToIdx[featName]] = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featName,
+						},
+						Value: types.Value{
+							Val: &types.Value_NullVal{
+								NullVal: types.Null_NULL,
+							},
+						},
+					}
+				}
+				return
+			}
+
+			scanner := iter.Scanner()
+			var entityKey string
+			var featureName string
+			var eventTs time.Time
+			var valueStr []byte
+			var deserializedValue types.Value
+			rowFeatures := make(map[string]FeatureData)
+			for scanner.Next() {
+				err := scanner.Scan(&entityKey, &featureName, &eventTs, &valueStr)
+				if err != nil {
+					errorsChannel <- errors.New("could not read row in query for (entity key, feature name, value, event ts)")
+					return
+				}
+				if err := proto.Unmarshal(valueStr, &deserializedValue); err != nil {
+					errorsChannel <- errors.New("error converting parsed Cassandra Value to types.Value")
+					return
+				}
+
+				if deserializedValue.Val != nil {
+					// Convert the value to a FeatureData struct
+					rowFeatures[featureName] = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featureName,
+						},
+						Timestamp: timestamppb.Timestamp{Seconds: eventTs.Unix(), Nanos: int32(eventTs.Nanosecond())},
+						Value: types.Value{
+							Val: deserializedValue.Val,
+						},
+					}
+				} else {
+					// Return FeatureData with a null value
+					rowFeatures[featureName] = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featureName,
+						},
+						Value: types.Value{
+							Val: &types.Value_NullVal{
+								NullVal: types.Null_NULL,
+							},
+						},
+					}
+				}
+			}
+
+			if err := scanner.Err(); err != nil {
+				errorsChannel <- errors.New("failed to scan features: " + err.Error())
```

Comment:
Can we modify the error message to have more context when you look at it?

fmt.Errorf("failed to scan features for entity key '%s' ......)

Reply:
In this part we only have access to a serialized entity key. It's going to look something like this: `0200000064656d6f5f6b657904000000080000000000000000000000`, I don't think it's useful to include it as the customer won't know what it is (they input a normal key, for the sample above it was `0`).